app/locations/middleware.js

```diff
@@ -15,12 +17,36 @@ function checkLocationsLength(req, res, next) {
 
 function setLocation(req, res, next) {
   const { locationId } = req.params
+
   const location = find(req.userLocations, { id: locationId })
 
   req.session.currentLocation = location
   next()
 }
 
+async function setRegion(req, res, next) {
+  const { regionId } = req.params
+
+  req.session.currentLocation = null
+
+  if (!regionId) {
+    req.session.currentRegion = null
+    next()
+    return
+  }
+
+  try {
+    const regions = await referenceDataService.getRegions()
+    req.session.currentRegion = find(regions, { id: regionId }) ?? null
+  } catch (error) {
+    req.session.currentRegion = null
+    next(new Error('Failed to retrieve the regions'))
```

Comment:
Could it be better to pass the error to next so that the stacktrace will contain a bit more information about the failure?